src/mlpack/methods/ann/layer/batch_norm.hpp

```diff
@@ -44,39 +47,81 @@ namespace ann /** Artificial Neural Network. */ {
+   * e.g., we apply batch normalization to the first 3 dimensions with number 
+   * of channels equal to 3rd dimension, and higher dimensions are left 
+   * untouched.  As an example, if we have a 3-dimensional input (call the 
+   * three dimensions rows, columns and slices), and `minAxis` & `maxAxis` is 
+   * 2, then we apply the same normalization across different slices.
+   */
+  BatchNormType();
 
   /**
-   * Create the BatchNorm layer object for a specified number of input units.
+   * Create the BatchNorm layer object for a specified axis of input units as 
+   * channels.
+   * 
+   * As an example, if we have a 3-dimensional input (call the three dimensions
+   * rows, columns and slices), and `minAxis` is 1 & `maxAxis` is 2, then the 
+   * number of channels is equal to `columns * slices`.
    *
-   * @param size The number of input units / channels.
+   * @param minAxis The min axis along which BatchNorm is applied. Before that,
+   *                it will be treated as input point.
+   * @param maxAxis The max axis along which BatchNorm is applied. After that,
+   *                it will be treated as another higher dimension point.
    * @param eps The epsilon added to variance to ensure numerical stability.
    * @param average Boolean to determine whether cumulative average is used for
    *                updating the parameters or momentum is used.
    * @param momentum Parameter used to to update the running mean and variance.
    */
-  BatchNorm(const size_t size,
-            const double eps = 1e-8,
-            const bool average = true,
-            const double momentum = 0.1);
+  BatchNormType(const size_t minAxis,
+                const size_t maxAxis,
+                const double eps = 1e-8,
+                const bool average = true,
+                const double momentum = 0.1);
+  
+  virtual ~BatchNormType() { }
+
+  //! Clone the BatchNormType object. This handles polymorphism correctly.
+  BatchNormType* Clone() const { return new BatchNormType(*this); }
+
+  //! Copy the other BatchNorm layer (but not weights).
+  BatchNormType(const BatchNormType& layer);
+
+  //! Take ownership of the members of the other BatchNorm layer (but not weights).
+  BatchNormType(BatchNormType&& layer);
+
+  //! Copy the other BatchNorm layer (but not weights).
+  BatchNormType& operator=(const BatchNormType& layer);
+
+  //! Take ownership of the members of the other BatchNorm layer (but not weights).
+  BatchNormType& operator=(BatchNormType&& layer);
 
   /**
    * Reset the layer parameters.
    */
-  void SetWeights(typename OutputType::elem_type* weightsPtr);
+  void SetWeights(typename MatType::elem_type* weightsPtr);
+
+  /**
+   * Initialize the weight matrix of the layer.
+   *
+   * @param W Weight matrix to initialize.
+   * @param rows Number of rows.
+   * @param * (cols) Number of columns.
+   */
+  void CustomInitialize(
+    MatType& W,
+    const size_t rows, 
+    const size_t /* cols */);
```

Comment:
```suggestion
      MatType& W,
      const size_t rows, 
      const size_t /* cols */);
```
Just a little style fix. :+1: